Add test for Method#to_proc with one and three arguments when calling Hash#each

marshalium opened the pull request:
This tests a bug that was found in JRuby: jruby/jruby#5896

This passes on MRI 2.7.2 but fails on JRuby 9.2.13.0. I've opened a PR to fix JRuby here: https://github.com/jruby/jruby/pull/6451

```
marshall@lappy7 spec % git rev-parse HEAD
3d552358e3544339096b99f645d414a6bf062f2e
```
```
marshall@lappy7 spec % RBENV_VERSION=2.7.2 ../mspec/bin/mspec  core/method/to_proc_spec.rb         
$ ruby /Users/marshall/src/ruby/mspec/bin/mspec-run core/method/to_proc_spec.rb
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
[- | ==================100%================== | 00:00:00]      0F      0E 

Finished in 0.003202 seconds

1 file, 12 examples, 40 expectations, 0 failures, 0 errors, 0 tagged
```
```
marshall@lappy7 spec % RBENV_VERSION=jruby-9.2.13.0 ../mspec/bin/mspec  core/method/to_proc_spec.rb
$ ruby /Users/marshall/src/ruby/mspec/bin/mspec-run core/method/to_proc_spec.rb
uri:classloader:/jruby/kernel/kernel.rb:4: warning: unsupported exec option: close_others
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 11.0.8+11 on 11.0.8+11 +jit [darwin-x86_64]
                                                                                             
1)
Method#to_proc returns a proc that properly handles when one or more items are yielded ERROR
ArgumentError: wrong number of arguments (given 1, expected 2)
/Users/marshall/src/ruby/spec/core/method/to_proc_spec.rb:113:in `two_params'
org/jruby/RubyHash.java:1415:in `each'
org/jruby/RubyEnumerable.java:886:in `map'
/Users/marshall/src/ruby/spec/core/method/to_proc_spec.rb:113:in `block in <main>'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyEnumerable.java:1809:in `all?'
org/jruby/RubyArray.java:1809:in `each'
/Users/marshall/src/ruby/spec/core/method/to_proc_spec.rb:4:in `<main>'
org/jruby/RubyKernel.java:1009:in `load'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyArray.java:1809:in `each'
[\ | ==================100%================== | 00:00:00]      0F      1E 

Finished in 0.053561 seconds

1 file, 12 examples, 38 expectations, 0 failures, 1 error, 0 tagged
```

marshalium commented:
The macos/2.5.8 ci failure looks like an unrelated build issue.

To be safe, I've confirmed that the spec I added passes locally on macos/2.5.8:
```
marshall@lappy7 spec % RBENV_VERSION=2.5.8 ../mspec/bin/mspec  core/method/to_proc_spec.rb
$ ruby /Users/marshall/src/ruby/mspec/bin/mspec-run core/method/to_proc_spec.rb
ruby 2.5.8p224 (2020-03-31 revision 67882) [x86_64-darwin19]
[- | ==================100%================== | 00:00:00]      0F      0E 

Finished in 0.005718 seconds

1 file, 12 examples, 40 expectations, 0 failures, 0 errors, 0 tagged      
```

eregon commented:
Actually I believe this is special behavior from Hash#{each,each_pair,map} (and not about Method#to_proc), which is tested in
https://github.com/ruby/spec/blob/b164536460f23127f3d2a496a93ea82275c7002a/core/hash/shared/each.rb#L24-L39
and
https://github.com/ruby/spec/blob/b164536460f23127f3d2a496a93ea82275c7002a/core/enumerable/shared/collect.rb#L54-L65

Note the first one (Hash#each) doesn't work on Ruby 3, and the second one (Enumerable#map for Hash) is being discussed in https://bugs.ruby-lang.org/issues/17197.

Could you check if those specs fail on JRuby?

Those files don't seem to test with a Method#to_proc with a single argument, that would be valuable to add and should work on all Ruby versions. Could you add a spec about that in each of these files? Testing with 3 arguments seems nice too.

marshalium commented:
Thanks for the quick review and pointing me at the specs I missed!

Yes, those existing specs do fail on JRuby and replicate the case I was testing for (I guess those specs are skipped in jruby ci):
```
marshall@lappy7 spec % RBENV_VERSION=jruby-9.2.13.0 ../mspec/bin/mspec core/enumerable/collect_spec.rb core/hash/each_spec.rb 
$ ruby /Users/marshall/src/ruby/mspec/bin/mspec-run core/enumerable/collect_spec.rb core/hash/each_spec.rb
uri:classloader:/jruby/kernel/kernel.rb:4: warning: unsupported exec option: close_others
jruby 9.2.13.0 (2.5.7) 2020-08-03 9a89c94bcc OpenJDK 64-Bit Server VM 11.0.8+11 on 11.0.8+11 +jit [darwin-x86_64]
                                                                                             
1)
Enumerable#collect yields 2 arguments for a Hash ERROR
ArgumentError: wrong number of arguments (given 1, expected 2)
/Users/marshall/src/ruby/spec/core/enumerable/shared/collect.rb:63:in `register'
org/jruby/RubyHash.java:1415:in `each'
org/jruby/RubyEnumerable.java:886:in `map'
/Users/marshall/src/ruby/spec/core/enumerable/shared/collect.rb:63:in `block in <main>'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyEnumerable.java:1809:in `all?'
org/jruby/RubyArray.java:1809:in `each'
/Users/marshall/src/ruby/spec/core/enumerable/collect_spec.rb:5:in `<main>'
org/jruby/RubyKernel.java:1009:in `load'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyArray.java:1809:in `each'
                                                                                             
2)
Hash#each yields 2 values and not an Array of 2 elements when given a callable of arity 2 ERROR
ArgumentError: wrong number of arguments (given 1, expected 2)
/Users/marshall/src/ruby/spec/core/hash/shared/each.rb:32:in `foo'
org/jruby/RubyHash.java:1415:in `each'
/Users/marshall/src/ruby/spec/core/hash/shared/each.rb:32:in `block in <main>'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyEnumerable.java:1809:in `all?'
org/jruby/RubyArray.java:1809:in `each'
/Users/marshall/src/ruby/spec/core/hash/each_spec.rb:7:in `<main>'
org/jruby/RubyKernel.java:1009:in `load'
org/jruby/RubyBasicObject.java:2694:in `instance_exec'
org/jruby/RubyArray.java:1809:in `each'
[| | ==================100%================== | 00:00:00]      0F      2E 

Finished in 0.092159 seconds

2 files, 19 examples, 29 expectations, 0 failures, 2 errors, 0 tagged
```

And you're correct that this is Hash specific behavior and not a general feature of yielding multiple things to blocks. The ruby source shows an explicitly arity check: https://github.com/ruby/ruby/blob/d1ba5545513b68d39ca29b578a42bd8d48a7804e/hash.c#L3092

I'll come back to this in a bit and add one and three param cases to those specs you pointed me at.

eregon commented:
@marshalium Could you come back to this when you get some time? :)

marshalium commented:
Yes! Sorry for the delay. I totally forgot to return to this. Will do this week!

marshalium commented:
I've updated this with the new specs as requested. 

The MacOS 2.6.6 failed on CI but it's already failing in the [same way on master](https://github.com/ruby/spec/runs/1464256764) so I think it's unrelated to my changes.

I also confirmed that the specs I modified still pass on a ruby 3.0.0 master build from today.

eregon reviewed (approved):
Thanks!
